src/libcrun/ebpf.c

```diff
@@ -0,0 +1,360 @@
+      .off   = 0,                               \
+      .imm   = 0 })
+
+# define BPF_EXIT_INSN()                        \
+  ((struct bpf_insn) {                          \
+    .code  = BPF_JMP | BPF_EXIT,                \
+      .dst_reg = 0,                             \
+      .src_reg = 0,                             \
+      .off   = 0,                               \
+      .imm   = 0 })
+#endif
+
+#ifdef HAVE_EBPF
+static size_t
+bpf_program_instructions (struct bpf_program *program)
+{
+  return program->used / (sizeof (struct bpf_insn));
+}
+#endif
+
+struct bpf_program *
+bpf_program_new (size_t size)
+{
+  struct bpf_program *p = xmalloc (size + sizeof (struct bpf_program));
+
+  p->used = 0;
+  p->private = 0;
+  p->allocated = size;
+
+  return p;
+}
+
+void
+bpf_program_append (struct bpf_program *p, void *data, size_t size)
+{
+  if (p->allocated <= p->used + size)
+    {
+      p->allocated += size * 2;
+      p = xrealloc (p, p->allocated + sizeof (struct bpf_program));
+    }
+  memcpy (p->program + p->used, data, size);
+  p->used += size;
+}
+
+int
+bpf_program_init_dev (struct bpf_program *program, libcrun_error_t *err)
+{
+#ifdef HAVE_EBPF
+  /* taken from systemd.  */
+  struct bpf_insn pre_insn[] = {
+                                /* type -> R2.  */
+                                BPF_LDX_MEM (BPF_H, BPF_REG_2, BPF_REG_1, 0),
+
+                                /* access -> R3.  */
+                                BPF_LDX_MEM (BPF_W, BPF_REG_3, BPF_REG_1, 0),
+                                BPF_ALU32_IMM (BPF_RSH, BPF_REG_3, 16),
+
+                                /* major -> R4.  */
+                                BPF_LDX_MEM (BPF_W, BPF_REG_4, BPF_REG_1, 4),
+
+                                /* minor -> R5.  */
+                                BPF_LDX_MEM (BPF_W, BPF_REG_5, BPF_REG_1, 8),
+  };
+  bpf_program_append (program, pre_insn, sizeof (pre_insn));
+#endif
+  return 0;
+}
+
+int
+bpf_program_append_dev (struct bpf_program *program, const char *access, char type, int major, int minor, bool accept, libcrun_error_t *err)
+{
+#ifdef HAVE_EBPF
+  int i;
+  int bpf_access = 0;
+  int bpf_type = type == 'b' ? BPF_DEVCG_DEV_BLOCK : BPF_DEVCG_DEV_CHAR;
+  bool has_type = type != 'a';
+  bool has_major = major >= 0;
+  bool has_minor = minor >= 0;
+  bool has_access = false;
+  int number_instructions = 0;
+  struct bpf_insn accept_block[] = {
+                                    BPF_MOV64_IMM (BPF_REG_0, accept ? 1 : 0),
+                                    BPF_EXIT_INSN (),
+  };
+
+  if (program->private & HAS_WILDCARD)
+    return 0;
+
+  for (i = 0; access[i]; i++)
+    {
+      switch (access[i])
+        {
+        case 'r':
+          bpf_access |= BPF_DEVCG_ACC_READ;
+          break;
+
+        case 'w':
+          bpf_access |= BPF_DEVCG_ACC_WRITE;
+          break;
+
+        case 'm':
+          bpf_access |= BPF_DEVCG_ACC_MKNOD;
+          break;
+        }
+    }
+
+  /*
```

Comment:
Why is this commented out block here?

Reply:
I've used it as a comment to explain what the eBPF assembly below does.

Reply:
Is this based on systemd implementation?

Reply:
loosely.  I've looked at the systemd implementation, copied the idea of moving the data into registers (that is the assembly in `bpf_program_init_dev`) then rewrote the rest as an exercise to get more familiar with ebpf.